.github/actions/publish/main.js

```diff
@@ -1,33 +1,72 @@
 const execSync = require('child_process').execSync
 const core = require('@actions/core')
+const fs = require('fs')
+const utils = require('../utils')
 
 /**
  * @fileoverview Publish to NPM, if the version is in the beta format
  * (e.g. 1.2.3-beta.4 or 1.2.3-alpha.1 etc.) then publish to a beta channel.
  */
 
-const version = core.getInput('version', {required: true})
-const dryRun = core.getBooleanInput('dryRun')
+/**
+ * @typedef {'normal'|'beta'} VersionType
+ * @typedef {{ type: VersionType, label?: string }} VersionInfo
+ */
+
+/**
+ * @param {string} version version string
+ * @return {VersionInfo|null} version info or null if invalid
+ */
+function parseVersion(version) {
+  const regexBetaVersion = /^\d+\.\d+\.\d+-(\w+)\.\d+$/;
+  const regexVersion = /^\d+\.\d+\.\d+$/;
 
-const regexBetaVersion = /^\d+\.\d+\.\d+-(\w+)\.\d+$/;
-const regexVersion = /^\d+\.\d+\.\d+$/;
+  const betaMatch = version.match(regexBetaVersion);
+  if (betaMatch) {
+    const label = betaMatch[1];
+    return { type: 'beta', label };
+  }
 
-const betaMatch = version.match(regexBetaVersion);
+  const isNormal = regexVersion.test(version);
+  if (isNormal) {
+    return { type: 'normal' };
+  }
 
-if (betaMatch) {
-  const betaLabel = betaMatch[1];
-  core.info(`Publishing a Beta version ${version}.`)
-  execSync(`npm publish ${dryRun ? '--dry-run' : ''} --tag ${betaLabel}`, { stdio: 'inherit' })
-  process.exit(0);
+  return null;
 }
 
-const isNormal = regexVersion.test(version);
 
-if (isNormal) {
-  core.info(`Publishing version ${version}.`)
-  execSync(`npm publish ${dryRun ? '--dry-run' : ''}`, { stdio: 'inherit' })
-  process.exit(0);
+const version = core.getInput('version', {required: true})
+const npmToken = core.getInput('npm-token', {required: true})
+const dryRun = core.getBooleanInput('dry-run')
+
+const versionInfo = parseVersion(version)
+
+if (versionInfo === null) {
+  core.error(`Invalid version format. Version: ${version}`)
+  process.exit(1)
+}
+
+const { type, label } = versionInfo;
+const args = ['--provenance']
+if (dryRun) {
+  args.push('--dry-run')
+}
+if (label) {
+  args.push(`--tag ${label}`)
 }
 
-core.error(`Invalid version format. Version: ${version}`);
-process.exit(1);
+core.info(`Publishing ${type === 'beta' ? 'Beta version' : 'version'} ${version}.`)
+
+const npmRcContent = `
+@castlabs:registry=https://registry.npmjs.org
+//registry.npmjs.org/:_authToken=\${NPM_TOKEN}
+`
+
+try {
+  fs.writeFileSync(utils.file('.npmrc'), npmRcContent)
+  execSync(`export NPM_TOKEN=${npmToken}; npm publish ${args.join(' ')}`, { stdio: 'inherit' })
```

Comment:
I think you do not even need to `export` the token here. I think you can either _pass_ this for that one shell commend, i.e. `NPM_TOKEN="" npm publish...` or (and I tend to like that more) pass it as an `env` option in the `options` of `execSync`.

Reply:
Nice, I will use `env`.